gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/FormatStep.java

```diff
@@ -0,0 +1,195 @@
+
+    @Override
+    protected Traverser.Admin<String> processNextStart() {
+        final Map<String, Object> values = new HashMap<>();
+
+        final Traverser.Admin traverser = this.starts.next();
+
+        boolean productive = true;
+        for (final String var : variables) {
+            final Object current = traverser.get();
+            // try to get property value
+            if (current instanceof Element) {
+                final Property prop = ((Element) current).property(var);
+                if (prop != null && prop.isPresent()) {
+                    values.put(var, prop.value());
+                    continue;
+                }
+            }
+
+            final TraversalProduct product =
+                    TraversalUtil.produce((S) this.getNullableScopeValue(Pop.last, var, traverser), this.traversalRing.next());
+
+            if (!product.isProductive() || product.get() == null) {
+                productive = false;
+                break;
+            }
+
+            values.put(var, product.get());
+        }
+        this.traversalRing.reset();
+
+        return productive ?
+                PathProcessor.processTraverserPathLabels(traverser.split(evaluate(values), this), this.keepLabels) :
+                EmptyTraverser.instance();
+    }
+
+    @Override
+    public String toString() {
+        return StringFactory.stepString(this, this.format, this.traversalRing);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = super.hashCode();
+        return Objects.hash(result, format, traversalRing);
+    }
+
+    @Override
+    public List<Traversal.Admin<S, String>> getLocalChildren() {
+        return this.traversalRing.getTraversals();
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        this.traversalRing.reset();
+    }
+
+    @Override
+    public FormatStep<S> clone() {
+        final FormatStep<S> clone = (FormatStep<S>) super.clone();
+        clone.format = this.format;
+        clone.variables = this.variables;
+        clone.traversalRing = this.traversalRing;
+        return clone;
+    }
+
+    @Override
+    public void setTraversal(final Traversal.Admin<?, ?> parentTraversal) {
+        super.setTraversal(parentTraversal);
+        this.traversalRing.getTraversals().forEach(this::integrateChild);
+    }
+
+    @Override
+    public Set<TraverserRequirement> getRequirements() {
+        return this.getSelfAndChildRequirements(TraverserRequirement.OBJECT, TraverserRequirement.SIDE_EFFECTS);
+    }
+
+    @Override
+    public Set<String> getScopeKeys() {
+        return variables;
+    }
+
+    @Override
+    public void setKeepLabels(final Set<String> labels) {
+        this.keepLabels = labels;
+    }
+
+    @Override
+    public Set<String> getKeepLabels() {
+        return this.keepLabels;
+    }
+
+    // private methods
+
+    private static final Pattern VARIABLE_PATTERN = Pattern.compile("%\\{(.*?)\\}");
+
+    Set<String> getVariables() {
```

Comment:
Nit: Did you want this to be package private? If not, explicitly add the private modifier.